### ISSUE TITLE
Avoid sending allocated data through graph

### ISSUE THREAD

**Issue by mrocklin** [StanfordLegion/task-bench#39]
https://github.com/StanfordLegion/task-bench/blob/1bf295870561a4f15ed613be6c31d4cd9cdc0023/dask/task_bench.py#L122-L142

It looks like you allocate a numpy array locally, and then put that numpy array into the task graph.  This will possibly send a lot of data through the centralized scheduler, which is likely to be suboptimal.

If the scratch variables are sizable, then you might want to delay those `np.empty` instead.  Perhaps something like this

```diff
-     scratch = [np.empty(graph.scratch_bytes_per_task, dtype=np.ubyte) for _ in range(graph.max_width)] 
+     scratch = [dask.delayed(np.empty)(graph.scratch_bytes_per_task, dtype=np.ubyte) for _ in range(graph.max_width)] 
```

This will cause them to be allocated on the workers as proper tasks rather than allocated locally and then communicating that data through the scheduler.

**Comment by mrocklin** [comment]
I could also be reading things incorrectly.  I've only glanced through things so far.

**Comment by elliottslaughter** [comment]
Yes, I think this is correct. I will apply this change and see how it impacts performance. (I'm also going to pull over the change I made in `task_bench_direct.py` to skip the scratch space when it's not needed.)

Thanks for pointing this out.

**Comment by elliottslaughter** [comment]
Should be fixed in e4e1bb15879dddebc80d948ec5f0eef3e0f210e2.

**Comment by mrocklin** [comment]
Thanks @elliottslaughter
